cache/cacheleveldown/leveldown.go

```diff
@@ -59,55 +63,66 @@ func (w *wrappedCache) Close() error {
 		return nil
 	}
 
+	w.manager.muCaches.Lock()
+
 	w.closed = true
 	err := w.wrappedCache.Close()
 	delete(w.manager.caches, w.id)
+
+	w.manager.muCaches.Unlock()
+
 	return err
 }
 
-func (l *levelDownCache) Load(directory string, dbAddress address.Address) (datastore.Datastore, error) {
-	cachePath := datastoreKey(directory, dbAddress)
-	var ds datastore.Datastore
+func (l *levelDownCache) Load(directory string, dbAddress address.Address) (ds datastore.Datastore, err error) {
+	keyPath := datastoreKey(directory, dbAddress)
 
-	if c, ok := l.caches[cachePath]; ok {
-		return c, nil
-	}
+	l.muCaches.Lock()
+	defer l.muCaches.Unlock()
 
-	logger().Debug(fmt.Sprintf("opening cache db from path %s", cachePath))
+	var ok bool
+	if ds, ok = l.caches[keyPath]; ok {
+		return
+	}
 
-	ds, err := leveldb.NewDatastore(cachePath, nil)
+	logger().Debug("opening cache db", zap.String("path", keyPath))
 
-	//ds = datastore.NewLogDatastore(ds, "cache-ds")
+	if directory == InMemoryDirectory {
+		ds, err = leveldb.NewDatastore("", nil)
+	} else {
+		ds, err = leveldb.NewDatastore(keyPath, nil)
+	}
 
 	if err != nil {
-		return nil, errors.Wrap(err, "unable to init leveldb datastore")
+		err = errors.Wrap(err, "unable to init leveldb datastore")
+		return
 	}
 
-	l.caches[cachePath] = &wrappedCache{wrappedCache: ds, id: cachePath, manager: l}
+	l.caches[keyPath] = &wrappedCache{wrappedCache: ds, id: keyPath, manager: l}
 
-	return ds, nil
+	return
 }
 
 func (l *levelDownCache) Close() error {
-	for k, c := range l.caches {
+	for _, c := range l.caches {
 		_ = c.Close()
-		delete(l.caches, k)
```

Comment:
because i noticed it was already removed from:
```go
func (w *wrappedCache) Close() error {
	if w.closed {
		return nil
	}

	w.manager.muCaches.Lock()

	w.closed = true
	err := w.wrappedCache.Close()
	delete(w.manager.caches, w.id)

	w.manager.muCaches.Unlock()

	return err
}
```